clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java

```diff
@@ -808,6 +808,162 @@ public void testIdempotenceWithMultipleInflights() throws Exception {
+        assertEquals(1, client.inFlightRequestCount());
+
+        // Later the second request returns successfully (which implies the first request
+        // was also successful). A successful result should be returned to the user.
+        sendIdempotentProducerResponse(1, tp0, Errors.NONE, 1L);
+        sender.runOnce();
+        assertTrue(request2.isDone());
+        assertEquals(1L, request2.get().offset());
+
+        // The producer epoch should not have been bumped and the next sequence
+        // number should be 2.
+        assertEquals(OptionalInt.of(1), transactionManager.lastAckedSequence(tp0));
+        assertEquals(new ProducerIdAndEpoch(producerId, (short) 0), transactionManager.producerIdAndEpoch());
+        appendToAccumulator(tp0);
+        sender.runOnce();
+        assertEquals(3, transactionManager.sequenceNumber(tp0).longValue());
+    }
+
+    @Test
+    public void testIdempotenceWithRequestTimeoutFollowingDeliveryTimeout() throws Exception {
+        final long producerId = 343434L;
+        TransactionManager transactionManager = createTransactionManager();
+        setupWithTransactionState(transactionManager);
+        prepareAndReceiveInitProducerId(producerId, Errors.NONE);
+        assertTrue(transactionManager.hasProducerId());
+
+        // Send two produce requests with a time gap between them. This allows
+        // us to test delivery timeout expiration of the first request without
+        // affecting the second.
+        Future<RecordMetadata> request1 = appendToAccumulator(tp0);
+        sender.runOnce();
+        assertEquals(1, transactionManager.sequenceNumber(tp0).longValue());
+        assertEquals(1, client.inFlightRequestCount());
+
+        time.sleep(DELIVERY_TIMEOUT_MS / 2);
+
+        Future<RecordMetadata> request2 = appendToAccumulator(tp0);
+        sender.runOnce();
+        assertEquals(2, transactionManager.sequenceNumber(tp0).longValue());
+        assertEquals(2, client.inFlightRequestCount());
+
+        // Both requests fail with a timeout error. The first one has reached the delivery
+        // timeout and should be failed. The second should be retried.
+        time.sleep(DELIVERY_TIMEOUT_MS / 2);
+        sendIdempotentProducerResponse(0, tp0, Errors.REQUEST_TIMED_OUT, -1L);
+        sendIdempotentProducerResponse(1, tp0, Errors.REQUEST_TIMED_OUT, -1L);
+        sender.runOnce();
+        TestUtils.assertFutureThrows(request1, TimeoutException.class);
+        assertFalse(request2.isDone());
+        assertEquals(0, client.inFlightRequestCount());
+
+        // Because we don't know the outcome of these requests yet, we should
+        // continue retrying until we know whether the requests were successful.
+        sender.runOnce();
+        assertEquals(1, client.inFlightRequestCount());
+
+        // Finally the response returns successfully and the result is returned
+        // to the user.
+        sendIdempotentProducerResponse(1, tp0, Errors.NONE, 1L);
```

Comment:
Is it possible for this to return successfully if the first one failed? I'm assuming we don't bump the epoch so I would expect out of sequence error (unless the first request was the first one to the partition)

Reply:
Yes, it is possible that the first request succeeded. We just didn't wait long enough to find that out.

Reply:
Hmmm. I guess I am missing how this is different from the first test. Is it because the second message also times out on 885?

Reply:
Yes, right. I wanted to verify here that the request is retried with the same sequence.